TF Work for survey results on object names

HelenBurge opened the pull request:
I have updated the rule in accessibility support and background to clarify what happens with no name, how it relates to the role attribute, added a failed example (might need tweaks) for an image in an object, and In the background, mention that when the object resource is not loaded the fallback content is rendered.

Closes issue(s):

- N/A

Need for Call for Review:
This will require a 1 week Call for Review 

---

## Pull Request Etiquette

### **When creating PR:**

- [ ] Make sure you're requesting to **pull a branch** (right side) to the `develop` branch (left side).
- [ ] Make sure you do not remove the "How to Review and Approve" section in your pull request description

### **After creating PR:**

- [ ] Add yourself (and co-authors) as "Assignees" for PR.
- [ ] Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- [ ] [Link the PR to any issue it solves](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue). This will be done automatically by [referencing the issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) at the top of this comment in the indicated place.
- [ ] Optionally request feedback from anyone in particular by assigning them as "Reviewers".

### **When merging a PR:**

- [ ] Close any issue that the PR resolves. This will happen automatically upon merging if the PR was correctly linked to the issue, e.g. by referencing the issue at the top of this comment.

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Call for Review period. In case of disagreement, the longer period wins.

HelenBurge committed:
TF Work for survey results on object names

carlosapaduarte commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +38,11 @@ The `object` element is not rendered for presentational purposes. If the `object
 ## Background
 
-Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately.
+Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately. When the object resource is not loaded the fallback content is rendered. The only [accessible name][] that is announced by the screen reader is for an `object` that has valid `data` set. The `object` method of adding an [accessible name][] will depend on the r ...
```
Shouldn't this be presented as a list?

carlosapaduarte replied:
While Trevor's suggestion did improve the readability and so I fully support keeping it, it did not address my concern. 
I was literally talking about presentation. I believe this should be presented as a list instead of a paragraph. The first sentence is not related to the second. The second is not related to the third. The third and fourth are related (or, with Trevor's suggestion, they are merged into a single sentence). So, what I'm asking for is a list with three items, instead of a paragraph.

HelenBurge replied:
Made into a list - thanks for the suggestion!

carlosapaduarte commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -129,6 +129,14 @@ This `object` element which embeds an audio resource has an empty [accessible na
 
+#### Failed Example 5
+
+This `object` element has an empty [accessible name][] as the `img` attribute has not been set correctly.
```
The description needs to be clearer. What `img` attribute has not been set correctly?

tombrunet replied:
I don't believe that this example actually fails anything regarding WCAG, though it's a very strange example. The `object` element has no `data` attribute, so it falls back to the `img`. The `img` has no `src`, which causes browsers to visibly show the `alt` text. The `alt` is exposed to ATs and to sighted users. It's "broken", but not with regards to accessibility...

HelenBurge replied:
I think I might need help on making this work - I will raise it in our call today!

tbostic32 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +38,11 @@ The `object` element is not rendered for presentational purposes. If the `object
 ## Background
 
-Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately.
+Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately. When the object resource is not loaded the fallback content is rendered. The only [accessible name][] that is announced by the screen reader is for an `object` that has valid `data` set. The `object` method of adding an [accessible name][] will depend on the r ...
```
```suggestion
Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately. When the object resource is not loaded the fallback content is rendered. The [accessible name][] announced by the screen reader is for an `object` with a valid `data` attribute and will depend on the role of the `data` attribute.
```
I think this is a bit more readable and may fix Carlos concern?

HelenBurge committed:
Update _rules/object-has-acessible-name-8fc3b6.md

Co-authored-by: Trevor R. Bostic <32486143+tbostic32@users.noreply.github.com>

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

WilcoFiers commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +38,13 @@ The `object` element is not rendered for presentational purposes. If the `object
 ## Background
 
-Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately.
+- Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately. 
```
I'm not a fan of using a list here TBH.

HelenBurge committed:
Update _rules/object-has-acessible-name-8fc3b6.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

HelenBurge commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +38,13 @@ The `object` element is not rendered for presentational purposes. If the `object
-Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately.
+- Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately. 
+- When the object resource is not loaded the fallback content is rendered. 
+- The [accessible name][] announced by the screen reader is for an `object` with a valid `data` attribute and will depend on the role of the `data` attribute.
```
It does mean the role attribute - as I was trying to answer: "Clarify how this relates to the role attribute" from what I understand, the type of data in the object determines the way an accessible name can be defined.

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

WilcoFiers commented on _rules/object-has-acessible-name-8fc3b6.md:
`````diff
@@ -179,6 +191,16 @@ There is no `object` element.
+````html
+<object>
+	<img src="/test-assets/shared/w3c-logo.png" alt="W3C logo" />
+</object>
`````
```suggestion
</object>
```
```

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update _rules/object-has-acessible-name-8fc3b6.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +41,16 @@ The `object` element is not rendered for presentational purposes. If the `object
 ## Background
 
-Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately.
+This rule is testing that the following 'object' attributes are followed:
```
I don't think "attributes" is the correct word. The list does not really mention attributes (in HTML sense of the term) of the `object` element. The mix-up with the technical language is a bit weird.

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +41,16 @@ The `object` element is not rendered for presentational purposes. If the `object
-Testing that the [accessible name][] describes the purpose of the `object` element is not part of this rule and must be tested separately.
+This rule is testing that the following 'object' attributes are followed:
+
+- The purpose is in the [accessible name][] but is not part of this rule and must be tested separately; and 
```
It's a bit weird to says that "this rule is testing the following stuff" and then list something that "is not part of the rule" 🤔

HelenBurge replied:
Yeah - I could take that out of the list :) I was trying to find a grouping and failed miserably!

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +41,16 @@ The `object` element is not rendered for presentational purposes. If the `object
+This rule is testing that the following 'object' attributes are followed:
+
+- The purpose is in the [accessible name][] but is not part of this rule and must be tested separately; and 
+- When the object resource is not loaded the fallback content is rendered; and
```
I'm not sure how this actually impact the rule.

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +41,16 @@ The `object` element is not rendered for presentational purposes. If the `object
+
+- The purpose is in the [accessible name][] but is not part of this rule and must be tested separately; and 
+- When the object resource is not loaded the fallback content is rendered; and
+- Valid `data` defines the [accessible name][] that is announced by the screen reader; and
```
I don't think that the `data` attribute of the `object` element plays a role when computing its accessible name. So, I do not understand what you mean by "`data` **defines** the accessible name".

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,11 +41,16 @@ The `object` element is not rendered for presentational purposes. If the `object
+- The purpose is in the [accessible name][] but is not part of this rule and must be tested separately; and 
+- When the object resource is not loaded the fallback content is rendered; and
+- Valid `data` defines the [accessible name][] that is announced by the screen reader; and
+- The [accessible name][] is defined depending on the role of the `data` attribute.
```
I'm also not sure that the (semantic) role of the `object` element actually has an impact on how its accessible name is computed.

HelenBurge replied:
If the object is an image, then the accessible name can be computed from an alt, but a MIME might use an aria-label. I could add that in as an example to clarify?

HelenBurge committed:
Update _rules/object-has-acessible-name-8fc3b6.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

HelenBurge committed:
Update _rules/object-has-acessible-name-8fc3b6.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

Amended the background to clarify points from Jean-Yves and remove the list structure as not really related data points!

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,12 +41,20 @@ The `object` element is not rendered for presentational purposes. If the `object
+
+When the object resource is not loaded the fallback content is rendered as shown in Inapplicable Example 7.
+
+Valid `data` attributes define the `object` [accessible name][] that is announced by the screen reader.  The [accessible name][] is defined depending on the type of the `data` attribute. For example,  an image will use an `alt` for the [accessible name][] but a MIME will not.
```
I'm still not fully understanding the details here. Especially what "MIME" means here. MIME types are a standard to declare type of data, and `image` is a standard MIME type.
So `<object data="foo.jpg" />` would have a MIME type of `image/jpeg`. Does this use `alt` for its accessible name?

Do you mean that "`image` MIME types uses `alt`, but other MIME types don't"?

Can you add test cases with this? It seems there is no test case with a `alt` attribute.

tbostic32 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,12 +41,20 @@ The `object` element is not rendered for presentational purposes. If the `object
+
+When the object resource is not loaded the fallback content is rendered as shown in Inapplicable Example 7.
+
+Valid `data` attributes define the `object` [accessible name][] that is announced by the screen reader.  The [accessible name][] is defined depending on the type of the `data` attribute. For example,  an image will use an `alt` for the [accessible name][] but a MIME will not.
```
Is MIME supposed to be `audio` or `video` as examples here?

HelenBurge replied:
yes - made an update

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

Jym77 committed:
Merge branch 'develop' into HelenBurge-patch-4

tombrunet commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,12 +41,20 @@ The `object` element is not rendered for presentational purposes. If the `object
+
+`Object` elements without an accessible name are ignored by assistive technologies unless they have an [explicit role][].
+
+When the object resource is not loaded the fallback content is rendered as shown in Inapplicable Example 7.
```
Calling out an example by number feels prone to be missed if the example numbering changes. I'm okay with it, but how do we normally make sure these references don't break?

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
````diff
@@ -179,6 +200,17 @@ There is no `object` element.
+```html
+<object>
+	<img src="/test-assets/shared/w3c-logo.png" alt="W3C logo" />
+</object>
````
Also, the `data` attribute seems to be mandatory on `object` elements: https://html.spec.whatwg.org/multipage/iframe-embed-object.html#attr-object-data
I guess when it's not here browsers fix things by using the content as fallback, but I'm never too keen on having non-failed examples that are not valid code.

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update _rules/object-has-acessible-name-8fc3b6.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

Jym77 committed:
Merge branch 'develop' into HelenBurge-patch-4

Jym77 commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,12 +41,20 @@ The `object` element is not rendered for presentational purposes. If the `object
+
+When the object resource is not loaded the fallback content is rendered as shown in the Inapplicable Example: "This `object` element does not need an accessible name because it loads no image, audio, or video.".
+
+The [MIME type][] of the resource embedded in the `data` attribute impacts how the [accessible name][] of the `object` is computed. For example, `object` embedding [image MIME type][] may use their `alt` attribute to compute their [accessible name][], but `object` embedding [audio or video MIME types][] may not.
```
This actually seems to be browser dependent.
When trying
```html
<object data="./test-assets/shared/w3c-logo.png" alt="W3C logo"></object>
<object data="./test-assets/moon-audio/moon-speech.mp3" alt="Moon speech"></object>

<object data="./test-assets/shared/w3c-logo.png" alt="W3C logo" type="image/png"></object>
<object data="./test-assets/moon-audio/moon-speech.mp3" alt="Moon speech" type="image/png"></object>

<object data="./test-assets/shared/w3c-logo.png" alt="W3C logo" type="audio/mpeg"></object>
<object data="./test-assets/moon-audio/moon-speech.mp3" alt="Moon speech" type="audio/mpeg"></object>
```
Chrome does not give any of the objects a name (and NVDA on top of it does not announce it); but Firefox makes the images a `graphic` with name "W3C logo" (and NVDA exposes it as "graphic W3C logo"), while not giving a name to the audios.

---
So, it seems that the `type` attribute actually isn't used at this level. 
However, the use of `alt` for images is browser dependent, so this should be an Accessibility Support note rather than a Background note.
I can't find a normative reference for the naming of `object` element, so I can't be sure which one is right 😕 Would you like to open an issue on HTML AAM and ask their opinion?

HelenBurge replied:
Sure - and we will let this one sit for now? 

What is the best place to ask? https://github.com/w3c/html-aam/issues?

Jym77 replied:
Yes, I'm rather wait for confirmation of what is the "good" solution so we know what we are actually suppose to test 😄 

Their issues list is indeed the place to ask.

HelenBurge replied:
https://github.com/w3c/html-aam/issues/431

scottaohara replied:
`alt` isn't supported on the object element.  Seems if this is going to be called out, it should be noted that it is not an official naming mechanism.  

Seems a potential issue to file against HTML to consider support the use of `alt` on the element... at least if it is specifically attempting to load an image resource...

HelenBurge replied:
Thanks @scottaohara!

Jym77 replied:
@HelenBurge So, my take would be:
* Not use `alt` on `object` in any Passed or Inapplicable example.
* Remove this note from Background
* Add a note to Accessibility Support saying that some UA/AT combination use `alt` on `object` embedding MIME `image`, despite `alt` not being allowed on `object` + link to [HTML spec of attributes allowed on object](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#the-object-element:concept-element-attributes). And so element may pass/fail depending on the UA/AT combination. (so, more or less the same content as this note, but presented a bit differently)
* We could have a Failed example with an `<object alt="…">`.

I wasn't part of the TF discussion around that, so not sure exactly what the request to add this were. It may be worth taking it again to the TF for further discussion.

HelenBurge replied:
Thanks @Jym77 - I moved the section into the accessibility support - but yes - I plan to discuss this in the meeting in 5 minutes as know enough to know it might need removing totally!

scottaohara commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -26,7 +26,10 @@ acknowledgments:
+This rule applies to any `object` element for which all the following are true:
+
+- The `object` element is [included in the accessibility tree][]; and
+- The `object` element embeds a resource whose MIME type is either [image](https://mimesniff.spec.whatwg.org/#image-mime-type)), or [audio or video](https://mimesniff.spec.whatwg.org/#audio-or-video-mime-type).
```
delete extra closing )
```suggestion
- The `object` element embeds a resource whose MIME type is either [image](https://mimesniff.spec.whatwg.org/#image-mime-type), or [audio or video](https://mimesniff.spec.whatwg.org/#audio-or-video-mime-type).
```

HelenBurge committed:
Update _rules/object-has-acessible-name-8fc3b6.md

Co-authored-by: Scott O'Hara <scottaohara@users.noreply.github.com>

HelenBurge committed:
Update object-has-acessible-name-8fc3b6.md

Added a caveat and moved the background into Accessibility Support

WilcoFiers commented on _rules/object-has-acessible-name-8fc3b6.md:
````diff
@@ -129,6 +138,16 @@ This `object` element which embeds an audio resource has an empty [accessible na
+	<img src="/test-assets/shared/w3c-logo.png" alt="W3C logo" />
+</object>
+```
+
````
````suggestion

#### Failed Example 6

This `object` element has `alt` attribute, however this will not create an  [accessible name][]. Because of this the accessible name is empty.

```html
<object data="/test-assets/moon-audio/moon-speech.mp3" alt="JFK's Going to the moon speech"></object>
```

````

WilcoFiers committed:
Update _rules/object-has-acessible-name-8fc3b6.md

WilcoFiers commented on _rules/object-has-acessible-name-8fc3b6.md:
````diff
@@ -129,6 +138,24 @@ This `object` element which embeds an audio resource has an empty [accessible na
+This `object` element has `alt` attribute, however this will not create an  [accessible name][]. Because of this the accessible name is empty.
+
+```html
+<object data="/test-assets/moon-audio/moon-speech.mp3" alt="JFK's Going to the moon speech"></object>
````
```suggestion
<object data="/test-assets/moon-audio/moon-speech.mp3" alt="Moon speech"></object>
```

WilcoFiers committed:
Update _rules/object-has-acessible-name-8fc3b6.md

WilcoFiers committed:
Update object-has-acessible-name-8fc3b6.md

kengdoj commented on _rules/object-has-acessible-name-8fc3b6.md:
```diff
@@ -38,12 +42,18 @@ The `object` element is not rendered for presentational purposes. If the `object
+
+`Object` elements without an accessible name are ignored by assistive technologies unless they have an [explicit role][].
+
+When the object resource is not loaded the fallback content is rendered as shown in the Inapplicable Example: "This `object` element does not need an accessible name because it loads no image, audio, or video.".
```
```suggestion
When the object resource is not loaded, the fallback content is rendered as shown in the Inapplicable Example: "This `object` element does not need an accessible name because it loads no image, audio, or video."
```

kengdoj reviewed (approved):
minor punctuation edits on line 55

WilcoFiers committed:
Apply suggestions from code review

Co-authored-by: Kathy Eng <kengdoj@users.noreply.github.com>

WilcoFiers committed:
Merge branch 'develop' into HelenBurge-patch-4
